libpromises/evalfunction.c

```diff
@@ -1331,79 +1349,159 @@ static FnCallResult FnCallPackagesMatching(ARG_UNUSED EvalContext *ctx, ARG_UNUS
             {
-                Log(LOG_LEVEL_ERR,
-                    "Line %d from package inventory '%s' did not yield 4 elements: %s",
-                    linenumber, filename, line);
-                ++linenumber;
-                SeqDestroy(list);
-                free(line);
+                Log(LOG_LEVEL_WARNING, "Can not read installed packages database "
+                    "for '%s' package module.", pm_name);
                 continue;
             }
+            
+            Seq *packages_form_module = SeqStringFromString(buff, '\n');
+            free(buff);
+            
+            if (packages_form_module)
+            {
+                // Iterate over and see where match is.
+                for (int i = 0; i < SeqLength(packages_form_module); i++)
+                {
+                    char *line = SeqAt(packages_form_module, i);
+                    size_t new_line_size = strlen(line) + pm_name_size + 2; // we need coma and terminator
+                    char new_line[new_line_size];
+                    strcpy(new_line, line);
+                    strcat(new_line, ",");
+                    strcat(new_line, pm_name);
+                    
+                    if (!AddPackagesmatchingJsonLine(matcher, json, new_line))
+                    {
+                        break;
+                    }
+                }
+                SeqDestroy(packages_form_module);
+            }
+            else
+            {
+                 Log(LOG_LEVEL_WARNING, "Can not parse packages database for '%s' "
+                     "package module.", pm_name);
+                 
+            }
+        }
+        CloseDB(db_cached);
+    }
+    return true;
+}
 
-            JsonElement *line_obj = JsonObjectCreate(4);
-            JsonObjectAppendString(line_obj, "name",    SeqAt(list, 0));
-            JsonObjectAppendString(line_obj, "version", SeqAt(list, 1));
-            JsonObjectAppendString(line_obj, "arch",    SeqAt(list, 2));
-            JsonObjectAppendString(line_obj, "method",  SeqAt(list, 3));
-            SeqDestroy(list);
+static FnCallResult FnCallPackagesMatching(ARG_UNUSED EvalContext *ctx, ARG_UNUSED const Policy *policy, const FnCall *fp, const Rlist *finalargs)
+{
+    const bool installed_mode = (0 == strcmp(fp->name, "packagesmatching"));
+    pcre *matcher;
+    {
+        const char *regex_package = RlistScalarValue(finalargs);
+        const char *regex_version = RlistScalarValue(finalargs->next);
+        const char *regex_arch = RlistScalarValue(finalargs->next->next);
+        const char *regex_method = RlistScalarValue(finalargs->next->next->next);
+        char regex[CF_BUFSIZE];
 
-            JsonArrayAppendObject(json, line_obj);
+        // Here we will truncate the regex if the parameters add up to over CF_BUFSIZE
+        snprintf(regex, sizeof(regex), "^%s,%s,%s,%s$",
+                 regex_package, regex_version, regex_arch, regex_method);
```

Comment:
Perhaps put that fix in a commit only on master then? I think this behavior is clearly broken, so this will be a bugfix, not a behavior change IMHO. After all, the old method can give incorrect matches.